File: backend/services/backup_service.py
"""
Backup Service for managing application backups.
Handles backup creation, listing, restoration, and deletion.
"""

from datetime import datetime
from typing import List, Optional
import asyncio
from sqlalchemy.orm import Session
from models.database import Backup, App
from services.proxmox_service import ProxmoxService


class BackupService:
    """Service for managing application backups."""

    def __init__(self, db: Session, proxmox_service: Optional[ProxmoxService] = None):
        """
        Initialize backup service.

        Args:
            db: Database session
            proxmox_service: Optional Proxmox service instance (for testing)
        """
        self.db = db
        self.proxmox = proxmox_service or ProxmoxService()

    async def create_backup(
        self,
        app_id: str,
        storage: str = "local",
        compress: str = "zstd",
        mode: str = "snapshot"
    ) -> Backup:
        """
        Create a backup for an application.

        Args:
            app_id: Application ID
            storage: Proxmox storage name
            compress: Compression type (zstd, gzip, none)
            mode: Backup mode (snapshot, stop)

        Returns:
            Backup object

        Raises:
            ValueError: If app not found
            Exception: If Proxmox backup creation fails
        """
        # Get app
        app = self.db.query(App).filter(App.id == app_id).first()
        if not app:
            raise ValueError(f"App {app_id} not found")

        # Generate filename
        timestamp = datetime.utcnow().strftime("%Y_%m_%d-%H_%M_%S")
        extension = ".tar.zst" if compress == "zstd" else ".tar.gz" if compress == "gzip" else ".tar"
        filename = f"vzdump-lxc-{app.lxc_id}-{timestamp}{extension}"

        # Create backup record
        backup = Backup(
            app_id=app_id,
            filename=filename,
            storage_name=storage,
            backup_type="vzdump",
            status="creating"
        )
        self.db.add(backup)
        self.db.commit()
        self.db.refresh(backup)

        try:
            # Start Proxmox backup
            task_id = await self.proxmox.create_vzdump(
                node=app.node,
                vmid=app.lxc_id,
                storage=storage,
                compress=compress,
                mode=mode
            )

            # Start background polling for completion
            asyncio.create_task(self._poll_backup_completion(backup.id))

            return backup

        except Exception as e:
            # Mark backup as failed
            backup.status = "failed"
            backup.error_message = str(e)
            self.db.commit()
            raise

    async def list_backups_for_app(self, app_id: str) -> List[Backup]:
        """
        List all backups for an application.

        Args:
            app_id: Application ID

        Returns:
            List of backups ordered by creation date (newest first)
        """
        backups = self.db.query(Backup).filter(
            Backup.app_id == app_id
        ).order_by(Backup.created_at.desc()).all()

        return backups

    async def get_backup(self, backup_id: int) -> Backup:
        """
        Get a specific backup.

        Args:
            backup_id: Backup ID

        Returns:
            Backup object

        Raises:
            ValueError: If backup not found
        """
        backup = self.db.query(Backup).filter(Backup.id == backup_id).first()
        if not backup:
            raise ValueError(f"Backup {backup_id} not found")

        return backup

    async def restore_from_backup(self, backup_id: int) -> dict:
        """
        Restore an application from a backup.

        Args:
            backup_id: Backup ID

        Returns:
            Restore operation result

        Raises:
            ValueError: If backup not found or not available
        """
        # Get backup
        backup = await self.get_backup(backup_id)

        # Check backup is available
        if backup.status != "available":
            raise ValueError(f"Backup is not available for restore (status: {backup.status})")

        # Get app
        app = self.db.query(App).filter(App.id == backup.app_id).first()
        if not app:
            raise ValueError(f"App {backup.app_id} not found")

        # Update backup status
        backup.status = "restoring"
        self.db.commit()

        try:
            # Stop container
            await self.proxmox.stop_lxc(app.node, app.lxc_id)

            # Restore from backup
            restore_task = await self.proxmox.restore_backup(
                node=app.node,
                vmid=app.lxc_id,
                backup_file=backup.filename,
                storage=backup.storage_name
            )

            # Start container
            await self.proxmox.start_lxc(app.node, app.lxc_id)

            # Update backup status
            backup.status = "available"
            self.db.commit()

            return {
                "success": True,
                "message": "Restore completed successfully",
                "backup_id": backup_id,
                "app_id": app.id
            }

        except Exception as e:
            # Mark backup as available again
            backup.status = "available"
            backup.error_message = f"Restore failed: {str(e)}"
            self.db.commit()
            raise

    async def delete_backup(self, backup_id: int) -> dict:
        """
        Delete a backup.

        Args:
            backup_id: Backup ID

        Returns:
            Deletion result

        Raises:
            ValueError: If backup not found
        """
        # Get backup
        backup = await self.get_backup(backup_id)

        # Get app for node info
        app = self.db.query(App).filter(App.id == backup.app_id).first()
        if not app:
            raise ValueError(f"App {backup.app_id} not found")

        try:
            # Delete from Proxmox storage
            await self.proxmox.delete_backup(
                node=app.node,
                storage=backup.storage_name,
                backup_file=backup.filename
            )

            # Delete from database
            self.db.delete(backup)
            self.db.commit()

            return {
                "success": True,
                "message": "Backup deleted successfully",
                "backup_id": backup_id
            }

        except Exception as e:
            raise Exception(f"Failed to delete backup: {str(e)}")

    async def _poll_backup_completion(self, backup_id: int, timeout: int = 3600):
        """
        Poll for backup completion in the background.

        Args:
            backup_id: Backup ID to poll
            timeout: Maximum time to wait in seconds (default 1 hour)
        """
        start_time = datetime.utcnow().timestamp()
        poll_interval = 5  # seconds

        while True:
            await asyncio.sleep(poll_interval)

            # Get current backup state
            backup = self.db.query(Backup).filter(Backup.id == backup_id).first()
            if not backup or backup.status != "creating":
                break

            # Get app
            app = self.db.query(App).filter(App.id == backup.app_id).first()
            if not app:
                break

            try:
                # Check if backup exists in Proxmox
                backup_list = await self.proxmox.get_backup_list(app.node, app.lxc_id)

                # Find our backup
                for proxmox_backup in backup_list:
                    if backup.filename in proxmox_backup.get("volid", ""):
                        # Backup completed!
                        backup.status = "available"
                        backup.size_bytes = proxmox_backup.get("size", 0)
                        backup.completed_at = datetime.utcnow()
                        self.db.commit()
                        return

            except Exception as e:
                # Log error but continue polling
                print(f"Error polling backup {backup_id}: {e}")

            # Check timeout
            if datetime.utcnow().timestamp() - start_time > timeout:
                backup.status = "failed"
                backup.error_message = "Backup creation timeout"
                self.db.commit()
                break

File: backend/frontend/js/services/backupService.js
/**
 * Backup Service
 * 
 * Provides backup management functionality for applications.
 * Handles creation, restoration, deletion, and polling of backups.
 * 
 * Features:
 * - List backups for an app
 * - Create new backups (snapshot mode, zstd compression)
 * - Restore from backup
 * - Delete backups
 * - Poll for backup completion (5s interval)
 * - Backup modal management
 * 
 * @module backupService
 */

/**
 * Current app ID for backup modal
 * @private
 */
let currentBackupAppId = null;

/**
 * Backup polling interval
 * @private
 */
let backupPollingInterval = null;

/**
 * Get API base URL
 * @returns {string} API base URL
 */
function getAPIBase() {
    return window.API_BASE || '/api/v1';
}

/**
 * Get authFetch function
 * @returns {Function} Auth fetch function
 */
function getAuthFetch() {
    return window.authFetch || fetch;
}

/**
 * Show notification
 */
function notify(message, type = 'info') {
    if (window.showNotification) {
        window.showNotification(message, type);
    } else {
        console.log(`[${type}] ${message}`);
    }
}

/**
 * Format date
 */
function fmtDate(date) {
    return window.formatDate ? window.formatDate(date) : new Date(date).toLocaleString();
}

/**
 * Format size
 */
function fmtSize(bytes) {
    return window.formatSize ? window.formatSize(bytes) : `${(bytes / 1024 / 1024).toFixed(2)} MB`;
}

/**
 * Get status icon
 */
function getStatusIcon(status) {
    if (window.getStatusIcon) {
        return window.getStatusIcon(status);
    }
    const icons = {
        'available': '✅',
        'creating': '⏳',
        'failed': '❌',
        'restoring': '🔄'
    };
    return icons[status] || '❓';
}

/**
 * Initialize Lucide icons
 */
function initIcons() {
    if (window.lucide && window.lucide.createIcons) {
        setTimeout(() => window.lucide.createIcons(), 100);
    }
}

/**
 * Show backup management modal
 * Opens modal and loads backups for specified app
 * 
 * @param {string} appId - Application ID
 * @returns {Promise<void>}
 */
export async function showBackupModal(appId) {
    currentBackupAppId = appId;
    const authFetch = getAuthFetch();
    const apiBase = getAPIBase();

    try {
        // Get app details
        const response = await authFetch(`${apiBase}/apps/${appId}`);
        const app = await response.json();
        
        const appNameEl = document.getElementById('backup-app-name');
        if (appNameEl) {
            appNameEl.textContent = app.name;
        }

        // Load backups
        await listBackups(appId);

        // Show modal
        const modal = document.getElementById('backupModal');
        if (modal) {
            modal.style.display = 'flex';
        }

        // Refresh icons after modal content is added
        initIcons();
    } catch (error) {
        notify('Failed to load backup modal', 'error');
        console.error('Error showing backup modal:', error);
    }
}

/**
 * Hide backup management modal
 * Closes modal and stops polling
 */
export function hideBackupModal() {
    const modal = document.getElementById('backupModal');
    if (modal) {
        modal.style.display = 'none';
    }
    currentBackupAppId = null;

    // Stop polling
    stopBackupPolling();
}

/**
 * List backups for an application
 * Fetches and renders backup list with status
 * 
 * @param {string} appId - Application ID
 * @returns {Promise<Array>} Array of backups
 */
export async function listBackups(appId) {
    const authFetch = getAuthFetch();
    const apiBase = getAPIBase();

    try {
        const response = await authFetch(`${apiBase}/apps/${appId}/backups`);
        const data = await response.json();
        const backups = data.backups || [];

        const listEl = document.getElementById('backup-list');
        if (!listEl) return backups;

        if (backups.length === 0) {
            listEl.innerHTML = '<p class="empty-state">No backups yet. Create your first backup to protect your data.</p>';
            return backups;
        }

        listEl.innerHTML = backups.map(backup => `
            <div class="backup-item" data-backup-id="${backup.id}">
                <div class="backup-info">
                    <div class="backup-filename">${backup.filename}</div>
                    <div class="backup-meta">
                        <span class="backup-date">
                            <i data-lucide="calendar"></i>
                            ${fmtDate(backup.created_at)}
                        </span>
                        <span class="backup-size">
                            <i data-lucide="hard-drive"></i>
                            ${fmtSize(backup.size_bytes)}
                        </span>
                        <span class="backup-status status-${backup.status}">
                            ${getStatusIcon(backup.status)}
                            ${backup.status}
                        </span>
                    </div>
                    ${backup.error_message ? `<div class="backup-error">${backup.error_message}</div>` : ''}
                </div>
                <div class="backup-actions">
                    ${backup.status === 'available' ? `
                        <button class="btn btn-sm btn-secondary" onclick="window.backupService.restoreBackup('${appId}', ${backup.id})" title="Restore from this backup">
                            <i data-lucide="rotate-ccw"></i>
                            Restore
                        </button>
                    ` : ''}
                    <button class="btn btn-sm btn-danger" onclick="window.backupService.deleteBackup('${appId}', ${backup.id})" title="Delete this backup">
                        <i data-lucide="trash-2"></i>
                        Delete
                    </button>
                </div>
            </div>
        `).join('');

        // Refresh icons
        initIcons();

        // Start polling if there are creating backups
        const creatingBackups = backups.filter(b => b.status === 'creating');
        if (creatingBackups.length > 0) {
            startBackupPolling(appId);
        }

        return backups;

    } catch (error) {
        notify('Failed to load backups', 'error');
        console.error('Error loading backups:', error);
        return [];
    }
}

/**
 * Create a new backup
 * Initiates backup creation with snapshot mode and zstd compression
 * 
 * @param {string} appId - Application ID (optional if modal is open)
 * @returns {Promise<void>}
 */
export async function createBackup(appId = null) {
    const targetAppId = appId || currentBackupAppId;
    if (!targetAppId) {
        notify('No app selected for backup', 'error');
        return;
    }

    const authFetch = getAuthFetch();
    const apiBase = getAPIBase();

    try {
        notify('Creating backup...', 'info');

        await authFetch(`${apiBase}/apps/${targetAppId}/backups`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                storage: 'local',
                compress: 'zstd',
                mode: 'snapshot'
            })
        });

        notify('Backup creation started', 'success');

        // Reload backups
        await listBackups(targetAppId);

    } catch (error) {
        notify('Failed to create backup', 'error');
        console.error('Error creating backup:', error);
        throw error;
    }
}

/**
 * Restore from backup
 * Restores application state from specified backup
 * 
 * @param {string} appId - Application ID
 * @param {number} backupId - Backup ID
 * @returns {Promise<void>}
 */
export async function restoreBackup(appId, backupId) {
    if (!confirm('Are you sure you want to restore from this backup? This will replace the current application state.')) {
        return;
    }

    const authFetch = getAuthFetch();
    const apiBase = getAPIBase();

    try {
        notify('Restoring from backup...', 'info');

        await authFetch(`${apiBase}/apps/${appId}/backups/${backupId}/restore`, {
            method: 'POST'
        });

        notify('Restore completed successfully', 'success');
        hideBackupModal();

        // Refresh app list
        if (window.dataService && window.dataService.loadDeployedApps) {
            await window.dataService.loadDeployedApps();
        } else if (window.loadApps) {
            await window.loadApps();
        }

    } catch (error) {
        notify('Failed to restore backup', 'error');
        console.error('Error restoring backup:', error);
        throw error;
    }
}

/**
 * Delete a backup
 * Permanently deletes specified backup
 * 
 * @param {string} appId - Application ID
 * @param {number} backupId - Backup ID
 * @returns {Promise<void>}
 */
export async function deleteBackup(appId, backupId) {
    if (!confirm('Are you sure you want to delete this backup? This action cannot be undone.')) {
        return;
    }

    const authFetch = getAuthFetch();
    const apiBase = getAPIBase();

    try {
        await authFetch(`${apiBase}/apps/${appId}/backups/${backupId}`, {
            method: 'DELETE'
        });

        notify('Backup deleted successfully', 'success');

        // Reload backups
        await listBackups(appId);

    } catch (error) {
        notify('Failed to delete backup', 'error');
        console.error('Error deleting backup:', error);
        throw error;
    }
}

/**
 * Refresh backups list
 * Reloads backups for current app
 * 
 * @returns {Promise<void>}
 */
export async function refreshBackups() {
    if (!currentBackupAppId) {
        notify('No app selected', 'error');
        return;
    }
    
    notify('Refreshing backups...', 'info');
    await listBackups(currentBackupAppId);
    notify('Backups refreshed', 'success');
}

/**
 * Start polling for backup completion
 * Polls every 5 seconds until backups are complete
 * @private
 * 
 * @param {string} appId - Application ID
 */
function startBackupPolling(appId) {
    // Clear existing interval
    stopBackupPolling();

    // Poll every 5 seconds
    backupPollingInterval = setInterval(async () => {
        if (currentBackupAppId === appId) {
            await listBackups(appId);
        } else {
            stopBackupPolling();
        }
    }, 5000);
}

/**
 * Stop backup polling
 * @private
 */
function stopBackupPolling() {
    if (backupPollingInterval) {
        clearInterval(backupPollingInterval);
        backupPollingInterval = null;
    }
}

/**
 * Get current backup app ID
 * @returns {string|null} Current app ID or null
 */
export function getCurrentBackupAppId() {
    return currentBackupAppId;
}

/**
 * Check if polling is active
 * @returns {boolean} True if polling is active
 */
export function isPolling() {
    return backupPollingInterval !== null;
}

// Backward compatibility: Expose to window
if (typeof window !== 'undefined') {
    window.backupService = {
        showBackupModal,
        hideBackupModal,
        listBackups,
        createBackup,
        restoreBackup,
        deleteBackup,
        refreshBackups,
        getCurrentBackupAppId,
        isPolling
    };
}
